### ISSUE TITLE
Binding between AST block and output (DOM)

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor#1560]
I want to create a web editor, with plain text on one side and the rendering on the other side.
When I select line 142 on my text editor, I want to automatically scroll to corresponding line on the rendering pane.

To achieve this I think we need a binding between input and output.

Let's say that we can retrieve the corresponding AST block(s) for a specific line range on the document:

```ruby
blocks = document.blocks_in_range(142, 150)
```

Then we could read a unique identifier on an AST block and search that identifier on the HTML DOM:

```javascript
var blockId = block.id;
$('#' + blockId).scroll();
```

Maybe this can be achieve with a `PostProcessor` ? But we still need a way to retrieve blocks from a line range.

With this feature and partial rendering we could update part of the DOM. Instant live reload even with long document :boom:

**Comment by jxxcarlson** [comment]
This would be great!  

The editor in http://www.noteshare.io <http://www.noteshare.io/>  does this after a fashion but it tracks the two
panes inaccurately.  Please keep me posted on this.  I would like to adopt your code:-)

Jim

**Comment by Mogztter** [comment]
@jxxcarlson Ok I will but first I want @mojavelinux's opinion on how we should implement this feature. Maybe this belongs to the extension lab ?

**Comment by jxxcarlson** [comment]
Absolutely!

I think this would be a killer feature for Asciidoctor.  

Anything I am able to do to help I would be glad to do.

**Comment by jxxcarlson** [comment]
I think we should also consider an extension that will produce a table of contents in several forms

 (1) just HTML: Thus it would be a set of links into the main file.  This would be very user
for an editor, since you could then place the TOC in a separate scrollable side bar.

(2) A structure that has all the info needed to produce the  above and which can be
manipulated programmatically.  

Having (2) is what is important.  That is how the sidebar TOC's in http://www.noteshare.io/titlepage/456 <http://www.noteshare.io/titlepage/456>  
and  http://nslab.herokuapp.com/document/63 are generated, though the data structures
are different.  

The second example may take a minute to spin up — it is running on a free dyno at heroku;
It is re-engineering of noteshare in a new framework called lotus: http://lotusrb.org/guides/getting-started/.

It should not be hard to extract what is needed from the AST.  Maybe you have already done this?

Jim

**Comment by Mogztter** [comment]
I think we already have everything we need.

The [source line number is available on Block](https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_block.rb#L89) when `sourcemap` option is set to true (related [test case](https://github.com/asciidoctor/asciidoctor/blob/master/test/document_test.rb#L292)) and we can [set an id on Node](https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_node.rb#L23).

Basically we can retrieve a block from a source line number, get the node id and scroll! I will create a simple example with Asciidoctor.js to showcase this "feature".

**Comment by jxxcarlson** [comment]
Hi @Mogztter,  I wrote a little test to see if I could get the line number of the blocks -- ran into a little problem -- nil values of line numbers.  Do you see what I am missing here?

```
# Filename: test_lines.rb
#
# File to test retrieval of source line numbers
# from the AST of an Asciidc doc
#
# Usage: '$ ruby test_file.rb'

str =<<EOF
== Foo

ho ho ho

== Bar

ha ha ha
EOF

require 'asciidoctor'

doc = Asciidoctor.load str

puts "Source text:\n------------\n#{str}\n------------\n"

puts "Number of blocks in doc = #{doc.blocks.count}"
puts "Line number of first block = #{doc.blocks[0].lineno}"
puts "Line number of second block = #{doc.blocks[1].lineno}"


puts "\nRerunning example with source in file instead of memory:\n\n"

IO.write('foo1234.adoc', str)

doc = Asciidoctor.load_file 'foo1234.adoc'

puts "Number of blocks in doc = #{doc.blocks.count}"
puts "Line number of first block = #{doc.blocks[0].lineno}"
puts "Line number of second block = #{doc.blocks[1].lineno}"
```

**Comment by Mogztter** [comment]
@jxxcarlson You need to add `sourcemap` option

```ruby
Asciidoctor.load str, :sourcemap
```
(Not sure about the Ruby syntax)

BTW, I have a working example here: https://github.com/Mogztter/asciidoctor-fast-live-reload
Let me know what you think!

**Comment by jxxcarlson** [comment]
Hi @Mogztter , thanks so much for the tip on `sourcemap`.  I've started experimenting with it: see [Hubpress blog](http://home.noteshareblog.io/2015/12/01/Experiments-with-Asciidoctor.html).  I looked at your Opal code, but that will take me some time to understand, since I haven't worked with Opal before.

**Comment by mojavelinux** [comment]
On Thu, Nov 12, 2015 at 2:38 PM, Guillaume Grossetie <
notifications@github.com> wrote:

> You need to add sourcemap option
>
> Asciidoctor.load str, :sourcemap
>
>
I believe it's:

doc = Asciidoctor.load str, sourcemap: true, safe: :safe
puts doc.blocks[0].source_location.inspect

This definitely needs to be documented in the user manual (though probably
better suited for the planned developer manual).

-Dan


-- 
Dan Allen | @mojavelinux | http://google.com/profiles/dan.j.allen

**Comment by mojavelinux** [comment]
On Wed, Nov 11, 2015 at 4:24 AM, Guillaume Grossetie <
notifications@github.com> wrote:

> I think we already have everything we need.
>
> The source line number is available on Block
> <https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_block.rb#L89>
> when sourcemap option is set to true (related test case
> <https://github.com/asciidoctor/asciidoctor/blob/master/test/document_test.rb#L292>)
> and we can set an id on Node
> <https://github.com/asciidoctor/asciidoctor/blob/master/lib/asciidoctor/abstract_node.rb#L23>
> .
>
> Basically we can retrieve a block from a source line number, get the node
> id and scroll! I will create a simple example with Asciidoctor.js to
> showcase this "feature".
>
+1

If there's more information you need from core, we can add it. I developed
the current sourcemap behavior without having a concrete use case, so I'm
sure that putting it to the test will show us what we're missing.

To answer your initial question, I think this is a collaboration between
core and an extension. Core needs to expose the low-level information and
the extension should be used to provide a query-like interface over that
information so that an editor can ask questions like "what blocks are in
this range".

Keep in mind that extensions don't have to live in the lab indefinitely.
That's just the playground area. Once an extension becomes generally
useful, I think we should graduate it into its own repository that is
published as a gem (or npm module). If an extension seems to small for one
repository, we can always group related functionality.

Cheers,

-Dan

-- 
Dan Allen | @mojavelinux | http://google.com/profiles/dan.j.allen

**Comment by mojavelinux** [comment]
Keep in mind that AsciidocFX is also doing something similar. An
improvement in one tool should help the others.

**Comment by mojavelinux** [comment]
> blocks = document.blocks_in_range(142, 150)

It's possible that an extension contribute addition methods to built-in classes like this. That's one of the beauties of Ruby.

**Comment by ahus1** [comment]
As discussed in https://github.com/asciidoctor/asciidoctor-intellij-plugin/pull/130, this should be a core feature of asciidoc. My suggestion from a tool vendor's perspective:

**User Story:**
As a tool vendor I want to have source line and source file information present in the HTML output when the `sourcemap` attribute is set.

**Acceptance Criteria:**
On each HTML element representing a block a data-sourcemap-file and data-sourcemap-lineno should be present

**Sidenote:** it would be nice to have source code position for single table rows and list elements. Unfortunately today a complete table and a complete iteration is always one block.

Until this is implemented, I've adapted an approach from AsciidocFX that uses sourcemap attribute and a treeprocessor ([see here](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/resources/sourceline-treeprocessor.rb)). When you use this treeprocessor when rendering HTML, the roles will be in the output and can be parsed ([see here for a javascript that scrolls to a cursor position](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/src/main/resources/org/asciidoc/intellij/editor/javafx/scrollToElement.js)). Kudos to AsciidocFX for implementing this first.

**Comment by Mogztter** [comment]
> On each HTML element representing a block a data-sourcemap-file and data-sourcemap-lineno should be present

@mojavelinux Do you think we should integrate this feature into Asciidoctor core ? In my opinion, using a [data attributes](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes) is the best approach.

Having to use/create an extension or code something feels like a little bit too much for this simple feature. Moreover, adding two data attributes `data-sourcemap-file=foo.adoc` and `data-sourcemap-line=12` will not require a lot of work in the html5 converter.
What is you position on this subject ?
